core/src/main/java/io/opencensus/tags/TagKey.java

```diff
@@ -94,12 +80,112 @@
    * @throws IllegalArgumentException if the name is not valid.
    */
   // TODO(sebright): Make this public once we support types other than String.
-  static TagKey<Boolean> createBooleanKey(String name) {
+  static TagKeyBoolean createBooleanKey(String name) {
     checkArgument(StringUtil.isValid(name));
-    return new AutoValue_TagKey<Boolean>(name, TAG_BOOLEAN);
+    return TagKeyBoolean.create(name);
+  }
+
+  public abstract String getName();
+
+  /**
+   * Applies a function to the {@code TagKey} subclass. The function that is called depends on the
+   * type of the tag key. This is similar to the visitor pattern.
+   *
+   * @param stringFunction the function to call when the {@code TagKey} is a {@code TagKeyString}.
+   * @param longFunction the function to call when the {@code TagKey} is a {@code TagKeyLong}.
+   * @param booleanFunction the function to call when the {@code TagKey} is a {@code TagKeyBoolean}.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag key's type.
+   */
+  // TODO(sebright): Should we make this public in the first release?
```

Comment:
If you want to go with this pattern (visitor) then you kind of have to do it from the beginning and probably support all types from the beginning.